libs/tlmatcher/src/main/java/org/codice/util/tlmatcher/TimeLimitedMatcher.java

```diff
@@ -0,0 +1,234 @@
+   * @throws TimeoutException if the RegEx search runs longer than {@code duration}
+   */
+  public static TimeLimitedMatcher create(
+      Pattern pattern, CharSequence inputString, Duration duration) {
+    TLCharSequence tlCharSequence = new TLCharSequence(inputString, calculateEndTime(duration));
+
+    Matcher matcher = pattern.matcher(tlCharSequence);
+    boolean matches = matcher.matches();
+
+    // If we've successfully matched, we can un-restrict the CharSequence.
+    // This would not likely be an issue if it were left unrestricted, as the charAt() method
+    // should not be called after the matches() method has completed; however, this will account
+    // for any future implementation changes that might alter that behavior.
+    tlCharSequence.unrestrict();
+
+    return new TimeLimitedMatcher(matcher, matches);
+  }
+
+  private TimeLimitedMatcher(Matcher matcher, boolean matches) {
+    this.matcher = matcher;
+    this.matches = matches;
+  }
+
+  /**
+   * Returns the input subsequence matched by the previous match.
+   *
+   * @return The (possibly empty) subsequence matched by the previous match, in string form
+   * @see Matcher#group()
+   */
+  public String group() {
+    return matcher.group();
+  }
+
+  /**
+   * Returns the input subsequence captured by the given group during the previous match operation.
+   *
+   * @param group The index of a capturing group in this matcher's pattern
+   * @return The (possibly empty) subsequence captured by the group during the previous match, or
+   *     null if the group failed to match part of the input
+   * @see Matcher#group(int)
+   */
+  public String group(int group) {
+    return matcher.group(group);
+  }
+
+  /**
+   * Returns the input subsequence captured by the given named-capturing group during the previous
+   * match operation.
+   *
+   * @param name The name of a named-capturing group in this matcher's pattern
+   * @return The (possibly empty) subsequence captured by the named group during the previous match,
+   *     or null if the group failed to match part of the input
+   * @see Matcher#group(String)
+   */
+  public String group(String name) {
+    return matcher.group(name);
+  }
+
+  /**
+   * Returns the number of capturing groups in this matcher's pattern.
+   *
+   * @return The number of capturing groups in this matcher's pattern
+   * @see Matcher#groupCount()
+   */
+  public int groupCount() {
+    return matcher.groupCount();
+  }
+
+  /**
+   * True if the entire region matched against the pattern.
+   *
+   * @return true if the entire region matched against the pattern.
+   * @see Matcher#matches()
+   */
+  public boolean matches() {
+    return matches;
+  }
+
+  public String toString() {
+    return String.format("TimeLimitedMatcher of [%s]", matcher.toString());
+  }
+
+  private static long calculateEndTime(Duration duration) {
+    return System.currentTimeMillis() + duration.toMillis();
```

Comment:
❓ Any reason we're using longs instead of using `Instant`s and using the built in functions like `isAfter`?

Reply:
No real reason, other than trying to be as close to the metal as possible here to keep it performant.

Reply:
I investigated `Instant.isAfter` , it's pretty bare metal.

Reply:
I'll do a quick analysis and make the change if I agree.

Reply:
It does involve object creation and that will happen each time a character is retrieved from the string being matched.